source/common/upstream/subset_lb.cc

```diff
@@ -173,14 +176,29 @@ void SubsetLoadBalancer::processSubsets(
           }
           subsets_modified.emplace(entry);
 
-          HostPredicate predicate =
-              std::bind(&SubsetLoadBalancer::hostMatches, this, kvs, std::placeholders::_1);
+          if (entry->initialized()) {
+            update_cb(entry);
+          } else {
+            HostPredicate predicate =
+                std::bind(&SubsetLoadBalancer::hostMatches, this, kvs, std::placeholders::_1);
 
-          cb(entry, predicate, adding_hosts);
+            new_cb(entry, predicate, adding_hosts);
+          }
         }
       }
     }
   }
+
+  forEachSubset(subsets_, [&](LbSubsetEntryPtr entry) {
+    if (subsets_modified.find(entry) != subsets_modified.end()) {
+      // Already handled due to hosts being added or removed.
+      return;
+    }
+
+    if (entry->initialized() && entry->active()) {
+      update_cb(entry);
```

Comment:
Do we double invoke `update_cb` here potentially (here and above)?

Reply:
No, subsets_modified tracks all the subsets for which we've called either update_cb or new_cb.

Reply:
I should clarify, it tracks the subsets for which new_cb or update_cb was called in block above. In this block, we see each subset once and only need to ignore those that were already updated (or just created).